integration/update_field_test.go

```diff
@@ -0,0 +1,913 @@
+				err = collection.FindOne(ctx, tc.filter).Decode(&actual)
+				require.NoError(t, err)
+
+				AssertEqualDocuments(t, tc.result, actual)
+			})
+		}
+	})
+
+	t.Run("errors", func(t *testing.T) {
+		t.Parallel()
+
+		for name, tc := range map[string]struct {
+			filter bson.D
+			update bson.D
+			err    *mongo.WriteError
+			alt    string
+		}{
+			"IncOnDocument": {
+				filter: bson.D{{"_id", "document"}},
+				update: bson.D{{"$inc", bson.D{{"value", int32(1)}}}},
+				err: &mongo.WriteError{
+					Code: 14,
+					Message: `Cannot apply $inc to a value of non-numeric type. ` +
+						`{_id: "document"} has the field 'value' of non-numeric type object`,
+				},
+			},
+			"IncOnArray": {
+				filter: bson.D{{"_id", "array"}},
+				update: bson.D{{"$inc", bson.D{{"value", int32(1)}}}},
+				err: &mongo.WriteError{
+					Code: 14,
+					Message: `Cannot apply $inc to a value of non-numeric type. ` +
+						`{_id: "array"} has the field 'value' of non-numeric type array`,
+				},
+			},
+			"IncOnString": {
+				filter: bson.D{{"_id", "string"}},
+				update: bson.D{{"$inc", "string"}},
+				err: &mongo.WriteError{
+					Code: 9,
+					Message: `Modifiers operate on fields but we found type string instead.` +
+						` For example: {$mod: {<field>: ...}} not {$inc: "string"}`,
+				},
+				alt: "Modifiers operate on fields but we found another type instead",
+			},
+			"IncWithStringValue": {
+				filter: bson.D{{"_id", "string-inc-error"}},
+				update: bson.D{{"$inc", bson.D{{"value", "bad value"}}}},
+				err: &mongo.WriteError{
+					Code:    14,
+					Message: `Cannot increment with non-numeric argument: {value: "bad value"}`,
+				},
+			},
+			"DoubleIncOnNullValue": {
+				filter: bson.D{{"_id", "string-inc-error"}},
+				update: bson.D{{"$inc", bson.D{{"value", float64(1)}}}},
+				err: &mongo.WriteError{
+					Code: 14,
+					Message: `Cannot apply $inc to a value of non-numeric type. ` +
+						`{_id: "string-inc-error"} has the field 'value' of non-numeric type string`,
+				},
+			},
+			"IntIncOnNullValue": {
+				filter: bson.D{{"_id", "string-inc-error"}},
+				update: bson.D{{"$inc", bson.D{{"value", int32(1)}}}},
+				err: &mongo.WriteError{
+					Code: 14,
+					Message: `Cannot apply $inc to a value of non-numeric type. ` +
+						`{_id: "string-inc-error"} has the field 'value' of non-numeric type string`,
+				},
+			},
+			"LongIncOnNullValue": {
+				filter: bson.D{{"_id", "string-inc-error"}},
+				update: bson.D{{"$inc", bson.D{{"value", int64(1)}}}},
+				err: &mongo.WriteError{
+					Code: 14,
+					Message: `Cannot apply $inc to a value of non-numeric type. ` +
+						`{_id: "string-inc-error"} has the field 'value' of non-numeric type string`,
+				},
+			},
+		} {
+			name, tc := name, tc
+			t.Run(name, func(t *testing.T) {
+				t.Parallel()
+				ctx, collection := setup(t, shareddata.Composites)
+				_, err := collection.InsertMany(ctx, []any{
+					bson.D{{"_id", "string-inc-error"}, {"value", "foo"}},
+				})
```

Comment:
This test does not use shareddata.Scalars, but inserts own data – that is not that different from data in our shared set.
Instead, this test should use shareddata.Scalars and not insert own data.

Please check other tests too